File: IPWorks Encrypt Samples/Hash/java/hash.java
/*
 * IPWorks Encrypt 2022 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks Encrypt in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksencrypt
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import ipworksencrypt.*;

public class hash extends ConsoleDemo {

  public static void main(String[] args) {
    

    try {   
      Hash hash1 = new Hash();
      
      System.out.println("******************************************************************************");
      System.out.println("* This demo show how to hash a file or string using a variety of algorithms  *");
      System.out.println("******************************************************************************");      
      
      while (true) {
        System.out.println("\nSpecify an algorithm:");
        System.out.println("0)  SHA1");
        System.out.println("1)  SHA-224");
        System.out.println("2)  SHA-256");
        System.out.println("3)  SHA-384");
        System.out.println("4)  SHA-512");
        System.out.println("5)  MD2");
        System.out.println("6)  MD4");
        System.out.println("7)  MD5");
        System.out.println("8)  RIPEMD-160");
        System.out.println("9)  MD5SHA1");
        System.out.println("10) HMAC-MD5");
        System.out.println("11) HMAC-SHA1");
        System.out.println("12) HMAC-SHA224");
        System.out.println("13) HMAC-SHA256");
        System.out.println("14) HMAC-SHA384");
        System.out.println("15) HMAC-SHA512");
        System.out.println("16) HMAC-RIPEMD160");
        System.out.println("17) SHA-3-224");
        System.out.println("18) SHA-3-256");
        System.out.println("19) SHA-3-384");
        System.out.println("20) SHA-3-512");
        hash1.setAlgorithm(Integer.parseInt(prompt("Algorithm", ":", "0")));
        
        hash1.setEncodeHash(ask("\nHex Encode Hash Value") == 'y');
        
        System.out.println("\nSelect an input source:");
        System.out.println("0) String");
        System.out.println("1) File");        
        
        if (Integer.parseInt(prompt("Input Source", ":", "0")) == 0) { //String
          System.out.println("Enter the message. To end the message, enter \":q\" on a single line by itself.");
            String inputString = "";
            String temp = prompt("Input String", ":\n");
          while (true) {
            if (temp.equals(":q"))
              break;
            inputString += temp + "\n";
            temp = input();
          }
          hash1.setInputMessage(inputString);
        }
        else {
          hash1.setInputFile(prompt("Input File"));
        }
        
        System.out.println("\nComputing Hash....");
        hash1.computeHash();
        System.out.println("\nHash Value: " + new String(hash1.getHashValue()));
        
        if (ask("\nDo you wish to hash another message") != 'y')
          break;
      } 
    } catch (IPWorksEncryptException ex) {
      displayError(ex);
    } catch (Exception ex) {
      displayError(ex);
    }
  }
}


class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }

  static String prompt(String label, String punctuation, String defaultVal)
  {
	System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
	String response = input();
	if(response.equals(""))
		return defaultVal;
	else
		return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksEncryptException) {
      System.out.print(" (" + ((IPWorksEncryptException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}

File: IPWorks Encrypt Samples/Hash/c++/hash.cpp
/*
 * IPWorks Encrypt 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks Encrypt in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksencrypt
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <stdlib.h>
#include <cstring>
#include "../../include/ipworksencrypt.h"
#define LINE_LEN 100
#define MESSAGE_LEN 1024

int main(int argc, char **argv)
{
  
  int ret_code = 0;
  Hash hash1;
  char algorithm[LINE_LEN];
  char command[LINE_LEN];
  char encodeHash[LINE_LEN];
  char inputSource[LINE_LEN];
  char inputFile[LINE_LEN];
  char inputString[MESSAGE_LEN];

  printf("******************************************************************************\n");
  printf("* This demo show how to hash a file or string using a variety of algorithms  *\n");
  printf("******************************************************************************\n");

  while (1) 
  {
    printf("Specify an algorithm:\n");
    printf("0)  SHA1 \n");
    printf("1)  SHA-224 \n");
    printf("2)  SHA-256 \n");
    printf("3)  SHA-384 \n");
    printf("4)  SHA-512 \n");
    printf("5)  MD2 \n");
    printf("6)  MD4 \n");
    printf("7)  MD5 \n");
    printf("8)  RIPEMD-160 \n");
    printf("9)  MD5SHA1 \n");
    printf("10) HMAC-MD5 \n");
    printf("11) HMAC-SHA1 \n");
    printf("12) HMAC-SHA224 \n");
    printf("13) HMAC-SHA256 \n");
    printf("14) HMAC-SHA384 \n");
    printf("15) HMAC-SHA512 \n");
    printf("16) HMAC-RIPEMD160 \n");
    printf("17) SHA-3-224 \n");
    printf("18) SHA-3-256 \n");
    printf("19) SHA-3-384 \n");
    printf("20) SHA-3-512 \n");
    printf("\nAlgorithm: ");
    fgets(algorithm, LINE_LEN, stdin);
    algorithm[strlen(algorithm)-1] = '\0';
    hash1.SetAlgorithm(atoi(algorithm));
  
    printf("\nSelect if hash value should be hex encoded:\n");
    printf("0) False\n");
    printf("1) True\n");
    printf("Hex Encode Hash Value: ");
    fgets(encodeHash, LINE_LEN, stdin);
    encodeHash[strlen(encodeHash)-1] = '\0';
    hash1.SetEncodeHash(atoi(encodeHash));

    printf("\nSelect an input source:\n");
    printf("0) String\n");
    printf("1) File\n");
    printf("\nInput Source: ");
    fgets(inputSource, LINE_LEN, stdin);
    inputSource[strlen(inputSource)-1] = '\0';

    int input = atoi(inputSource);
    if (input == 0) //String 
    {    
      printf("Enter the message. To end the message, enter \":q\" on a single line by itself.\n");
      printf("\nInput String:\n");
      inputString[0] = '\0';
      while(fgets(command, LINE_LEN, stdin))
      {
        if (strcmp(command, ":q\n") == 0)
          break;
        strcat(inputString, command);
      }
      hash1.SetInputMessage(inputString, strlen(inputString));
    }
    else 
    {
      printf("Input File: ");
      fgets(inputFile, LINE_LEN, stdin);
      inputFile[strlen(inputFile)-1] = '\0';
      hash1.SetInputFile(inputFile);
    }
    
    printf("\nComputing Hash....\n");
    ret_code = hash1.ComputeHash();
    if (ret_code) goto done;
    char *hashValue;
    int len;
    hash1.GetHashValue(hashValue, len);
    printf("\nHash Value: %s\n", hashValue);
    
    printf("\nDo you wish to hash another message (y/n)? ");
    fgets(command, LINE_LEN, stdin);
    command[strlen(command)-1] = '\0';
    if (strcmp("y", command) != 0 && strcmp("Y", command) != 0)
      break;
  }

done:
  if (ret_code) 
  {
    printf("Error %d: %s\n", hash1.GetLastErrorCode(), hash1.GetLastError());
  }

  fprintf(stderr, "\n\npress <return> to continue...");
  getchar();
  exit(ret_code);
  return 0;
}















